### PR TITLE
Pyrocar boss

### PR DESCRIPTION
**Pull request by Alex-Turtle**
Pyrocar is based on a saxxy entry, used back during the SLAG days for the Fight Night gamemode.

The boss is equipped with a regular backburner and posses the same speed as a medic. Pyrocar comes along with a "Bhop" rage and a float super jump.  It also adds a "Force foward" that forces the player to go foward.

Its superrage spawns a fully charged kritz medic.

### CODE REVIEW

File: addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+	"models/player/items/pyro/pyro_hat.mdl",
+	"models/player/items/pyro/fireman_helmet.mdl",
+	"models/player/items/pyro/pyro_chef_hat.mdl",
+	"models/player/items/all_class/ghostly_gibus_pyro.mdl",
+	"models/player/items/pyro/pyro_madame_dixie.mdl"
+	
+};
+
+static int g_iCosmetics[] =
+{
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp]: Lonely bracket

File: addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	"vsh_rewrite/pyrocar/pyrocar_transport.mp3"
+};
+
+static char g_strPyrocarKillBuilding[][] =  {
+	"vsh_rewrite/pyrocar/pyrocar_burning.mp3",
+	"vsh_rewrite/pyrocar/pyrocar_goingdown.mp3"
+};
+
+static char g_strPrecacheCosmetics[][] =
+{
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp]: Lonely bracket

File: addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	public void OnThink()
+	{
+		if (GameRules_GetRoundState() == RoundState_Preround) return;
+		if (!IsPlayerAlive(this.iClient)) return;
+		
+		int iTeam = GetClientTeam(this.iClient);
+		
+		if(g_bStompEnabled[this.iClient])
+		{
+			float vecExplosionPos[3], vecExplosionOrigin[3];
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp]: ``vecExplosionPos`` is pointless, only has one use in ``TF2_Explode`` below and is assigned ``vecExplosionOrigin`` anyway

File: addons/sourcemod/scripting/vsh/abilities/ability_float_jump.sp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	{
+		public get()
+		{
+			return g_flFloatJumpMaxDistance[this.iClient];
+		}
+		public set(float val)
+		{
+			g_flFloatJumpMaxDistance[this.iClient] = val;
+		}
+	}
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/abilities/ability_float_jump.sp]: These properties are never used

File: addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	{
+		public get()
+		{
+			return g_flRageHopMaxHeight[this.iClient];
+		}
+		public set(float val)
+		{
+			g_flRageHopMaxHeight[this.iClient] = val;
+		}
+	}
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp]: Property is never used

File: addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+};
+
+static char g_strPyrocarLastMan[][] =  {
+	"vsh_rewrite/pyrocar/pyrocar_transport.mp3"
+};
+
+static char g_strPyrocarKillBuilding[][] =  {
+	"vsh_rewrite/pyrocar/pyrocar_burning.mp3",
+	"vsh_rewrite/pyrocar/pyrocar_goingdown.mp3"
+};
```
Comment (Mikusch) [inline addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp]: You still need to add the sound files

File: addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			SetEntProp(this.iClient, Prop_Send, "m_bJumping", true);
+			
+			TeleportEntity(this.iClient, NULL_VECTOR, NULL_VECTOR, vecVel);
+			
+			g_bStompEnabled[this.iClient] = true;
+		}
+		else
+		{
+		int iWeapon = GetPlayerWeaponSlot(this.iClient, WeaponSlot_Primary);
+		TF2Attrib_SetByName(iWeapon, "damage force reduction", 0.5);
```
[the remark this answers is not in the document]
Comment (Mikusch) [reply addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp]: Somehow GitHub ate my comment, but use ``TF2Attrib_SetByDefIndex`` to stay consistent with other bosses

[the remark this answers is not in the document]
Comment (FortyTwoFortyTwo) [reply addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp]: Also may need valid entity check (`iWeapon > MaxClient`)

File: addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	
+	public void OnThink()
+	{
+		if (GameRules_GetRoundState() == RoundState_Preround) return;
+		if (!IsPlayerAlive(this.iClient)) return;
+		
+		int iTeam = GetClientTeam(this.iClient);
+		
+		if(g_bStompEnabled[this.iClient])
```
Comment (FortyTwoFortyTwo) [inline addons/sourcemod/scripting/vsh/abilities/ability_rage_hop.sp]: Space between if and bracket, several more below

File: addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+static char g_strPrecacheCosmetics[][] =  {
+	"models/player/items/pyro/pyro_hat.mdl",
+	"models/player/items/pyro/fireman_helmet.mdl",
+	"models/player/items/all_class/ghostly_gibus_pyro.mdl",
+	"models/player/items/pyro/pyro_madame_dixie.mdl",
+	"models/player/items/pyro/pyro_chef_hat.mdl"
+
```
Comment (FortyTwoFortyTwo) [inline addons/sourcemod/scripting/vsh/bosses/boss_pyrocar.sp]: Extra uneeded space

**Pull request merged**